state_scheduler.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"sort"
+	"time"
+
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+type Scheduler interface {
+	ScheduleState(state State, freq time.Duration, once bool) error
+	RemoveState(state State, freq time.Duration) error
+	Wait(cancel chan struct{}) (State, bool)
+}
+
+type stateScheduler struct {
+	events []*stateEvent
+}
+
+type byTime []*stateEvent
+
+func (e byTime) Len() int      { return len(e) }
+func (e byTime) Swap(i, j int) { e[i], e[j] = e[j], e[i] }
+func (e byTime) Less(i, j int) bool {
+	return e[i].runNext.Before(e[j].runNext)
+}
+
+type stateEvent struct {
+	every   time.Duration
+	state   State
+	runOnce bool
+
+	runNext time.Time
+	id      string
+}
+
+func NewStateScheduler() Scheduler {
+	return &stateScheduler{}
+}
+
+func getKey(t time.Duration, s State) string {
+	return string(t) + s.Id().String()
+}
+
+func (s *stateScheduler) ScheduleState(state State, freq time.Duration, once bool) error {
+	key := getKey(freq, state)
+
+	for _, e := range s.events {
+		if e.id == key {
+			return errors.Errorf("State %s with frequency %s already scheduled",
+				state.Id(), freq)
+		}
+	}
+
+	log.Debugf("Scheduling state %s with frequency %s", state.Id(), freq)
+
+	ev := &stateEvent{
+		every:   freq,
+		state:   state,
+		runOnce: once,
+
+		// calculate the next run time
+		runNext: time.Now().Add(freq),
+		id:      key,
+	}
+	s.events = append(s.events, ev)
+
+	return nil
+}
+
+func (s *stateScheduler) RemoveState(state State, freq time.Duration) error {
+	key := getKey(freq, state)
```

Comment:
Why use a key at all? The state is enough to identify the element in the list, since you only allow one of each state anyway. And having to pass both the state and the frequency into `RemoveState` is a somewhat strange API, I think the state alone should be enough.